### PR TITLE
inital changes to support pre provisioned switches and links

### PR DESCRIPTION
**Pull request by mthurstocisco**

### CODE REVIEW

File: roles/dtc/common/templates/ndfc_create_fabric.j2 [hunk 1]
```diff
@@ -94,6 +94,8 @@
 {# ------------------------------------------------------ #}
 {# Underlay General Parameters                            #}
 {# ------------------------------------------------------ #}
+  ENABLE_SGT: false
```
Comment (mikewiebe) [inline roles/dtc/common/templates/ndfc_create_fabric.j2:97]: Once https://github.com/netascode/ansible-dc-vxlan/pull/203 is merged we will need to pull in the changes to this PR to handle versioning between templates for properties like `ENABLE_SGT` and `ENABLE_PVLAN`

File: roles/dtc/common/templates/ndfc_inventory.j2 [hunk 2]
```diff
@@ -14,4 +14,15 @@
   max_hops: 0 # this is the default value as it is not defined into the data model
   role: {{ switch['role'] }}
   preserve_config: false
+{% if switch['poap'] is defined %}
```
Comment (mikewiebe) [inline roles/dtc/common/templates/ndfc_inventory.j2:17]: We will need to refactor this after https://github.com/netascode/ansible-dc-vxlan/pull/204 is merged

File: roles/dtc/create/tasks/sub_main.yml [hunk 3]
```diff
@@ -67,9 +66,17 @@
...
+
 - name: Manage NDFC Fabric Policies
   ansible.builtin.import_tasks: policies.yml
-  when: 
+  when:
     - (MD_Extended.vxlan.policy is defined) and (MD_Extended.vxlan.policy.policies | length > 0)
     - changes_detected_policy
-  tags: "{{ nac_tags.create_policy }}"
+  tags:
+    - "{{ nac_tags.create_policy[0] }}"
```
Comment (mikewiebe) [inline roles/dtc/create/tasks/sub_main.yml:82]: Why do we need the `[0]` here?

File: roles/dtc/create/tasks/link_loop.yml [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+# IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+# CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+#
+# SPDX-License-Identifier: MIT
+
+---
+- name: Add links to not_required_links if they already exist
+  ansible.builtin.set_fact:
+    not_required_links: "{{ not_required_links + [ link ]}}"
+  loop: "{{ result_links.response }}"
```
Comment (mikewiebe) [inline roles/dtc/create/tasks/link_loop.yml:26]: Let's try and move this into an action plugin instead of using the `loop` construct.  It should be significantly faster

Comment (mthurstocisco) [reply roles/dtc/create/tasks/link_loop.yml:26]: that should be done now, take a look and see what you think

File: roles/dtc/common/tasks/sub_main.yml [hunk 5]
```diff
@@ -20,7 +20,6 @@
 # SPDX-License-Identifier: MIT
 
 ---
-
```
[the remark this answers is not in the document]
Comment (mthurstocisco) [reply roles/dtc/common/tasks/sub_main.yml:23]: this was my linter i think, this isnt the only place its changed it for me :/ will go through and check

File: roles/dtc/common/templates/ndfc_fabric_links.j2 [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+#jinja2: trim_blocks: True, lstrip_blocks: False
+{% for link in MD_Extended.vxlan.topology.fabric_links %}
+  - dst_fabric: "{{ MD.vxlan.global.name }}"
+    src_interface: "{{ link.source_interface }}" 
+    dst_interface: "{{ link.dest_interface }}" 
+    src_device: "{{ link.source_device }}" 
+    dst_device: "{{ link.dest_device }}" 
+    template: int_pre_provision_intra_fabric_link
+    profile:
+        admin_state: true
```
[the remark this answers is not in the document]
Comment (mthurstocisco) [reply roles/dtc/common/templates/ndfc_fabric_links.j2]: done

File: roles/dtc/common/tasks/sub_main.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (mthurstocisco) [reply roles/dtc/common/tasks/sub_main.yml]: sorted

File: roles/dtc/deploy/tasks/sub_main.yml [hunk 7]
```diff
@@ -35,6 +35,7 @@
     method: POST
     path: "/appcenter/cisco/ndfc/api/v1/lan-fabric/rest/control/fabrics/{{ MD.vxlan.global.name }}/config-save"
   when: MD_Extended.vxlan.topology.switches | length > 0
+  ignore_errors: True
```
[the remark this answers is not in the document]
Comment (mthurstocisco) [reply roles/dtc/deploy/tasks/sub_main.yml:39]: We added some logic to look for specific responses from the endpoint but still require the ignore_errors

File: roles/dtc/common/templates/ndfc_create_fabric.j2 (hunk 1 above)
Comment (mthurstocisco) [reply roles/dtc/common/templates/ndfc_create_fabric.j2:116]: should be fixed now, updated from develop

File: roles/dtc/common/templates/ndfc_attach_vrfs_loopbacks.j2 [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+    "vrfName": "{{ vrf["name"] }}",
+    {% for switch in MD.vxlan.topology.switches %}
+      {% if switch.management.management_ipv4_address == attach["hostname"] %}
+    "serialNumber": "{{ switch.serial_number }}",
+      {% endif %}
+    {% endfor %}
+    "freeformConfig": "",
+    "extensionValues": "",
+    "vlan": "{{ vrf["vlan_id"] }}",
+    "deployment": true,
```
Comment (mtarking) [inline roles/dtc/common/templates/ndfc_attach_vrfs_loopbacks.j2:14]: @mthurstocisco this was the deployment question

Comment (mthurstocisco) [reply roles/dtc/common/templates/ndfc_attach_vrfs_loopbacks.j2]: updated, think this just got copied across from our old solution

File: roles/dtc/deploy/tasks/sub_main.yml (hunk 7 above)
[the remark this answers is not in the document]
Comment (mthurstocisco) [reply roles/dtc/deploy/tasks/sub_main.yml]: added the rescue block we talked about, puts message back to the user but carries on

File: roles/dtc/common/templates/ndfc_vpc_peering_serials.j2 [hunk 9]
```diff
@@ -0,0 +1,20 @@
+---
+# This NDFC vPC Peering config data structure is auto-generated
+# DO NOT EDIT MANUALLY
+#
+
+{% if MD.vxlan.topology.vpc_peers is defined and MD.vxlan.topology.vpc_peers is not none %}
```
[the remark this answers is not in the document]
Comment (mthurstocisco) [reply roles/dtc/common/templates/ndfc_vpc_peering_serials.j2]: Awaiting https://github.com/netascode/ansible-dc-vxlan/pull/233 being merged then will refactor

File: roles/dtc/common/templates/ndfc_inventory.j2 (hunk 2 above)
Comment (mthurstocisco) [reply roles/dtc/common/templates/ndfc_inventory.j2]: this should be done now too

File: roles/dtc/common/templates/ndfc_vpc_peering_serials.j2 (hunk 9 above)
[the remark this answers is not in the document]
Comment (mthurstocisco) [reply roles/dtc/common/templates/ndfc_vpc_peering_serials.j2]: file no longer required, fixed in DCNM collection

File: roles/dtc/create/tasks/vrf_loopbacks.yml [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+#
+# SPDX-License-Identifier: MIT
+
+---
+
+- name: Manage VRFs and Networks Entry Point
+  ansible.builtin.debug:
+    msg:
+      - "----------------------------------------------------------------"
+      - "+     Manage VRFs and Networks Fabric {{ MD.vxlan.global.name }}"
```
Comment (mikewiebe) [inline roles/dtc/create/tasks/vrf_loopbacks.yml:28]: Update this debug to indicate we are managing VRF loopbacks

File: roles/dtc/create/tasks/vrf_loopbacks.yml [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+  set_fact:
+    attach_payload: "{{ attach_payload | default([]) +  [{'vrfName':vrf.name, 'lanAttachList':vrf_loopback_config}]}}"
+
+- name: Call API to attach VRF to Leaf Switches
+  cisco.dcnm.dcnm_rest:
+    path: "/appcenter/cisco/ndfc/api/v1/lan-fabric/rest/top-down/v2/fabrics/{{ MD.vxlan.global.name }}/vrfs/attachments"
+    method: "POST"
+    json_data: "{{ attach_payload | to_json}}"
+
+- name: Clear payload
```
Comment (mikewiebe) [inline roles/dtc/create/tasks/vrf_loopbacks.yml:62]: Why is this needed?

File: roles/dtc/create/tasks/vrfs_networks.yml [hunk 12]
```diff
@@ -56,3 +55,16 @@
...
     - changes_detected_networks
+
+# --------------------------------------------------------------------
+# Manage Loopback VRF attachments on NDFC
+# --------------------------------------------------------------------
+- name: Attach VRF Loopbacks per VRF
+  include_tasks: vrf_loopbacks.yml
+  loop_control:
+    loop_var: vrf
+  loop: "{{ MD.vxlan.overlay_services.vrfs }}"
```
Comment (mikewiebe) [inline roles/dtc/create/tasks/vrfs_networks.yml:66]: Are we able to refactor this to not loop?

File: plugins/action/dtc/existing_links_check.py [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+
+from ansible.utils.display import Display
+from ansible.plugins.action import ActionBase
+
+display = Display()
+
+
+class ActionModule(ActionBase):
+
+    def run(self, tmp=None, task_vars=None):
```
[the remark this answers is not in the document]
Comment (mthurstocisco) [reply plugins/action/dtc/existing_links_check.py:35]: done

File: roles/dtc/create/tasks/vrfs_networks.yml (hunk 12 above)
Comment (mthurstocisco) [reply roles/dtc/create/tasks/vrfs_networks.yml:66]: removed, left over from an old looping version

File: roles/dtc/create/tasks/vrf_loopbacks.yml (hunk 11 above)
Comment (mthurstocisco) [reply roles/dtc/create/tasks/vrf_loopbacks.yml:62]: removed

File: plugins/action/dtc/existing_links_check.py [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+        results = super(ActionModule, self).run(tmp, task_vars)
+        existing_links = self._task.args['existing_links']['response']
```
[the remark this answers is not in the document]
Comment (mthurstocisco) [reply plugins/action/dtc/existing_links_check.py]: done, actually changed the logic around so that bit of python never gets called if the response from the previous play is empty.

File: roles/dtc/common/templates/ndfc_attach_vrfs_loopbacks.j2 [hunk 15]
```diff
@@ -0,0 +1,18 @@
+[
```
Comment (mikewiebe) [inline roles/dtc/common/templates/ndfc_attach_vrfs_loopbacks.j2:1]: Seeing a couple of issues here with develop merged in

- `attach_group` key needs to be updated to be `vrf_attach_group`
- I see the following error when I don't have any vrf loopback data defined

```bash
ansible.errors.AnsibleUndefinedVariable: 'dict object' has no attribute 'loopback_id'. 'dict object' has no attribute 'loopback_id'
failed: [nac-ndfc1 -> localhost] (item={'name': 'NaC-VRF04', 'vrf_id': 150004, 'vlan_id': 2004, 'vrf_attach_group': 'leaf3'}) => {
    "ansible_loop_var": "vrf",
    "changed": false,
    "msg": "AnsibleUndefinedVariable: 'dict object' has no attribute 'loopback_id'. 'dict object' has no attribute 'loopback_id'",
    "vrf": {
        "name": "NaC-VRF04",
        "vlan_id": 2004,
        "vrf_attach_group": "leaf3",
        "vrf_id": 150004
    }
}
```

Comment (mthurstocisco) [reply roles/dtc/common/templates/ndfc_attach_vrfs_loopbacks.j2:1]: hey man, gonna look at refactoring this loopback attach stuff today anyway to try and get rid of the loop per vrf, will sort the above while im there.

File: roles/dtc/create/tasks/vrfs_networks.yml (hunk 12 above)
Comment (mthurstocisco) [reply roles/dtc/create/tasks/vrfs_networks.yml]: should be fixed now

File: roles/dtc/common/templates/ndfc_attach_vrfs.j2 [hunk 16]
```diff
@@ -52,7 +52,6 @@
 {% for attach in MD_Extended.vxlan.overlay_services.vrf_attach_groups_dict[vrf['vrf_attach_group']] %}
     - ip_address: {{ attach['mgmt_ip_address'] }}
 {% endfor %}
-  deploy: false
 {% endif %}
-
+  deploy: false
```
Comment (mikewiebe) [inline roles/dtc/common/templates/ndfc_attach_vrfs.j2:56]: Spoke with Peter about this and this is probably ok.  Moves the `deploy` key outside of the attach check.

This:

```
{# ------------------------------------------------------ #}
{# Attach Group Section #}
{# ------------------------------------------------------ #}
{% if vrf['vrf_attach_group'] is defined %}
  attach:
{% for attach in MD_Extended.vxlan.overlay_services.vrf_attach_groups_dict[vrf['vrf_attach_group']] %}
    - ip_address: {{ attach['mgmt_ip_address'] }}
{% endfor %}
{% endif %}
  deploy: false
{% endfor %}
```
vs. this
```
{# ------------------------------------------------------ #}
{# Attach Group Section #}
{# ------------------------------------------------------ #}
{% if vrf['vrf_attach_group'] is defined %}
  attach:
{% for attach in MD_Extended.vxlan.overlay_services.vrf_attach_groups_dict[vrf['vrf_attach_group']] %}
    - ip_address: {{ attach['mgmt_ip_address'] }}
{% endfor %}
  deploy: false
{% endif %}
```

File: roles/dtc/common/templates/ndfc_attach_vrfs_loopbacks.j2 (hunk 15 above)
Comment (mthurstocisco) [reply roles/dtc/common/templates/ndfc_attach_vrfs_loopbacks.j2:1]: this has been resolved now too

File: roles/dtc/deploy/tasks/sub_main.yml [hunk 17]
```diff
@@ -30,12 +30,20 @@
...
+  block:
+    - name: Config-Save for Fabric {{ MD.vxlan.global.name }}
+      cisco.dcnm.dcnm_rest:
+        method: POST
+        path: "/appcenter/cisco/ndfc/api/v1/lan-fabric/rest/control/fabrics/{{ MD.vxlan.global.name }}/config-save"
+      when: MD_Extended.vxlan.topology.switches | length > 0
+      register: config_save
+      # TODO: Need to add logic to only save if changes are made
+  
+  rescue:
```
Comment (mikewiebe) [inline roles/dtc/deploy/tasks/sub_main.yml:43]: Is there a specific scenario that is failing for you?  I recall this use to just be an ignore errors but if we actually get a failure we are just printing the message and moving on.  Should we actually stop execution if we encounter a failure?

**Review by mikewiebe — APPROVED** [review]
